### DISCUSSION TITLE
How to avoid a keymgmt provider being fetched by default?

### DISCUSSION THREAD

**Discussion by matthauck** [openssl/openssl#22139] [Q&A] [answered]
Context: I'm working on a (built-in) keymgmt provider that provides sign and digest-sign operations for private keys whose key material is present in another system. 

However, it looks like openssl is loading my keymgmt provider too eagerly and trying to use it for verify operations using public keys that are unrelated to the aforementioned keymgmt provider. 

It seems ultimately to happen here:

https://github.com/openssl/openssl/blob/2743594d73e65c38375c619e89ec62579e2c24a9/crypto/evp/pmeth_lib.c#L276

This ends up calling `evp_pkey_export_to_provider` to import the public key into my provider, which does not work. 

I can get around this by specifying a `propquery` value on all EVP_PKEY contexts, but that approach feels error prone / whack-a-mole. It seems to me that there should be a way for a provider to not fetched for `EVP_KEYMGMT_fetch( globalCtx, ecKeyType, "" )`. 

The plot thickens further: if i call `EVP_KEYMGMT_fetch` explicitly on "EC" key types after loading my provider, but prior to performing any key operations, it also resolves the issue for me. Which seems to indicate that it is an issue of which provider is cached first perhaps internally? (I got lost trying to debug the store of cached providers inside openssl). This resolution also feels brittle / unsatisfactory.

Perhaps there is a way on the provider interface to indicate which keys are allowed to be imported that I missed? I have specified `OSSL_FUNC_KEYMGMT_IMPORT_TYPES` to return an empty array, but this doesn't seem to be a factor in the decision openssl is making. 

Hopefully I'm just missing something. Thanks for any help you can provide!

**Comment by paulidale** [comment]
EC keys currently cannot be distinguished by the curve they use.  It's a known shortcoming and something we will have to address at some point.

**Reply by matthauck** [reply]
Is there a github issue already for this? if not, would you like me to open one?

**Comment by mattcaswell** [comment]
> I can get around this by specifying a propquery value on all EVP_PKEY contexts,

Can you set a global propquery?

**Reply by matthauck** [reply]
Ah, with `EVP_set_default_properties`? I hadn't thought of that. 

I attempted that just now to set default properties to `provider=default`, however, and now signatures with my provider keys can't sign/decrypt anything with implicit fetching.

- I have an EVP_PKEY loaded with explicit properties. This worked like before.
- I call `EVP_SignFinal` (not the _ex version, want to rely on the keymgmt in the EVP_PKEY to choose the right provider)
- This seems to do all the right things until `evp_pkey_signature_init` ends up calling `evp_signature_fetch_from_prov` to fetch the signature op from my provider, but this appears to use the default properties, because it does not find my signature provider

**Reply by paulidale** [reply]
Suggest you set a default property of `?provider=default` which will prefer the default provider but not mandate it.

**Reply by matthauck** [reply]
Ah, right. Yeah, that makes more sense. 🤦

**Comment by matthauck** [comment, marked as answer]
Found a workaround: 
- Make a separate `OSSL_LIB_CTX`
- Load my provider into this context, not the global context
- Import my keys using this context explicitly

This seems to allow everything to work with implicit fetching.
